// react-bootstrap-intl-tel-input/dist/IntlTelInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _countryData = require('country-data');

var _countryData2 = _interopRequireDefault(_countryData);

var _googleLibphonenumber = require('google-libphonenumber');

var _escapeStringRegexp = require('escape-string-regexp');

var _escapeStringRegexp2 = _interopRequireDefault(_escapeStringRegexp);

var _FlagIcon = require('react-flag-kit/lib/FlagIcon');

var _FlagIcon2 = _interopRequireDefault(_FlagIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var IntlTelInput = function (_Component) {
  _inherits(IntlTelInput, _Component);

  function IntlTelInput() {
    _classCallCheck(this, IntlTelInput);

    var _this = _possibleConstructorReturn(this, (IntlTelInput.__proto__ || Object.getPrototypeOf(IntlTelInput)).call(this));

    _this.phoneUtil = _googleLibphonenumber.PhoneNumberUtil.getInstance();
    _this.countries = _countryData2.default.callingCountries.all.filter(function (country) {
      return country.status === 'assigned';
    });
    _this.mouseDownOnMenu = false;
    _this._pageClick = _this.pageClick.bind(_this);
    _this.missingFlags = { AQ: 'WW', BQ: 'NL', EH: 'WW-AFR', MF: 'FR', SH: 'GB' };
    _this.state = {
      open: false,
      selectedCountry: {},
      phoneNumber: '',
      inputWithoutCountryCode: '',
      searchTerm: '',
      valid: false,
      filteredCountries: [],
      preferredCountries: [],
      paginateCount: 1,
      multiSelectOpen: false,
      multiSelectItem: {},
      lastPreferred: '',
      tabbedIndex: 0
    };
    return _this;
  }

  _createClass(IntlTelInput, [{
    key: 'getPreferredCountries',
    value: function getPreferredCountries() {
      var preferredCountries = this.props.preferredCountries;

      if (preferredCountries && preferredCountries.length) {
        var _preferredCountries = preferredCountries.map(function (country) {
          return country.toUpperCase();
        });
        var preferred = this.countries.filter(function (country) {
          return _preferredCountries.indexOf(country.alpha2) !== -1;
        }).reverse();
        var regular = this.countries.filter(function (country) {
          return _preferredCountries.indexOf(country.alpha2) === -1;
        });
        var orderedCountries = preferred.concat(regular);
        this.setState({ preferredCountries: orderedCountries, lastPreferred: preferred[preferred.length - 1] });
        return orderedCountries;
      }
    }
  }, {
    key: 'mapErrorMessage',
    value: function mapErrorMessage(message) {
      var _props = this.props,
          minLengthMessage = _props.minLengthMessage,
          maxLengthMessage = _props.maxLengthMessage,
          callingCodeMessage = _props.callingCodeMessage,
          catchAllMessage = _props.catchAllMessage;

      if (message === 'The string supplied did not seem to be a phone number' || message === 'The string supplied is too short to be a phone number' || message === 'Phone number too short after IDD') {
        return minLengthMessage;
      } else if (message === 'The string supplied is too long to be a phone number') {
        return maxLengthMessage;
      } else if (message === 'Invalid country calling code') {
        return callingCodeMessage;
      } else {
        return catchAllMessage;
      }
    }
  }, {
    key: 'formatValidation',
    value: function formatValidation(valid, internalMessage, friendlyMessage, parsed) {
      return {
        valid: valid,
        internalMessage: internalMessage,
        friendlyMessage: friendlyMessage,
        parsed: parsed
      };
    }
  }, {
    key: 'validateNumber',
    value: function validateNumber(phoneNumber, alpha2) {
      if (alpha2) {
        var _alpha2 = alpha2 === 'unknown' ? '' : alpha2;
        try {
          this.phoneUtil.parse(phoneNumber, _alpha2);
        } catch (e) {
          var message = e.message;

          return this.formatValidation(false, message, this.mapErrorMessage(message), null);
        }
        var parsed = this.phoneUtil.parse(phoneNumber, _alpha2);
        var valid = this.phoneUtil.isPossibleNumber(parsed);
        return this.formatValidation(valid, '', valid ? '' : this.mapErrorMessage(), parsed);
      } else {
        var callingCodeMessage = this.props.callingCodeMessage;

        return this.formatValidation(false, '', callingCodeMessage, null);
      }
    }
  }, {
    key: 'onKeyDown',
    value: function onKeyDown(e) {
      var _this2 = this;

      var _state = this.state,
          tabbedIndex = _state.tabbedIndex,
          paginateCount = _state.paginateCount,
          open = _state.open,
          filteredCountries = _state.filteredCountries;

      if (open) {
        var dropdownItemHeight = parseInt(window.getComputedStyle(this.countryDropdown.children[0]).getPropertyValue('height'));
        var dropdownHeight = parseInt(window.getComputedStyle(this.countryDropdown).getPropertyValue('height'));
        var halfwayPoint = dropdownHeight / dropdownItemHeight / 2;
        var paginate = this.props.paginate;

        var key = e.keyCode;
        if (key === 27) {
          this.setState({ open: false, tabbedIndex: 0 });
        } else if (key === 40 | key === 9) {
          e.preventDefault();
          var newIndex = tabbedIndex === filteredCountries.length ? filteredCountries.length : tabbedIndex + 1;
          this.setState({ tabbedIndex: newIndex }, function () {
            _this2.countryDropdown.scrollTop = dropdownItemHeight * (newIndex - halfwayPoint);
            if (paginate && paginateCount && paginate * paginateCount === newIndex - 2) {
              _this2.setState({ paginateCount: paginateCount + 1 });
            }
          });
        } else if (key === 38) {
          var _newIndex = tabbedIndex === 0 ? 0 : tabbedIndex - 1;
          this.setState({ tabbedIndex: _newIndex }, function () {
            _this2.countryDropdown.scrollTop = dropdownItemHeight * (_newIndex - halfwayPoint);
          });
        } else if (key === 13) {
          var _filteredCountries = this.state.filteredCountries;

          this.selectCountry(_filteredCountries[tabbedIndex - 1]);
        }
      }
    }
  }, {
    key: 'lookupCountry',
    value: function lookupCountry(callingCode) {
      return callingCode === parseInt(1) ? _countryData2.default.countries.US : _countryData2.default.lookup.countries({ countryCallingCodes: '+' + callingCode }).filter(function (country) {
        return country.status === 'assigned';
      })[0];
    }
  }, {
    key: 'testNumber',
    value: function testNumber(number) {
      return new RegExp(/^[0-9]+$/).test(number);
    }
  }, {
    key: 'unformatNumber',
    value: function unformatNumber(number) {
      return number.replace(/[^0-9]/g, '');
    }
  }, {
    key: 'onChangePhone',
    value: function onChangePhone() {
      var value = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
      var mounted = arguments[1];
      var onChange = this.props.onChange;
      var _state2 = this.state,
          selectedCountry = _state2.selectedCountry,
          callingCode = _state2.callingCode;

      if (selectedCountry !== 'unknown') {
        var alpha2 = selectedCountry.alpha2,
            name = selectedCountry.name;

        var countryCode = alpha2.toUpperCase();
        var inputWithoutCountryCode = value.substring(callingCode.length + 1).trim();
        this.setState({ inputWithoutCountryCode: inputWithoutCountryCode });
        var unformattedNumber = this.unformatNumber(inputWithoutCountryCode);
        if (this.testNumber(unformattedNumber) || value === '' || inputWithoutCountryCode === '') {
          var formatter = new _googleLibphonenumber.AsYouTypeFormatter(countryCode);
          var formattedNumberArray = unformattedNumber.split('').map(function (char) {
            return formatter.inputDigit(char);
          });
          var phoneNumber = callingCode + ' ' + (formattedNumberArray.length ? formattedNumberArray[formattedNumberArray.length - 1] : unformattedNumber);
          var validation = this.validateNumber(phoneNumber, alpha2);
          this.setState({ phoneNumber: phoneNumber, validation: validation, inputWithoutCountryCode: inputWithoutCountryCode });
          if (onChange && !mounted) {
            onChange({ phoneNumber: unformattedNumber, callingCode: callingCode, inputWithoutCountryCode: this.unformatNumber(inputWithoutCountryCode), alpha2: alpha2, name: name, validation: validation });
          }
          formatter.clear();
        }
      } else if (this.testNumber(value) || value === '') {
        var _validation = this.validateNumber(value);
        var _unformattedNumber = this.unformatNumber(value);
        this.setState({ phoneNumber: value, inputWithoutCountryCode: value, validation: _validation });
        if (onChange && !mounted) {
          onChange({ phoneNumber: _unformattedNumber, inputWithoutCountryCode: _unformattedNumber, callingCode: null, name: null, alpha2: null, validation: _validation });
        }
      }
    }
  }, {
    key: 'getFormattedNumber',
    value: function getFormattedNumber(callingCode) {
      var input = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

      return callingCode + ' ' + input;
    }
  }, {
    key: 'cancelMultiSelect',
    value: function cancelMultiSelect() {
      var _this3 = this;

      this.setState({ multiSelectOpen: false, multiSelectItem: {} }, function () {
        _this3.multiSelect.style.zIndex = '-1';
      });
    }
  }, {
    key: 'onChangeTypeAhead',
    value: function onChangeTypeAhead(value) {
      var preferredCountries = this.state.preferredCountries;

      var filteredCountries = this.countries.filter(function (country) {
        var name = country.name,
            countryCallingCodes = country.countryCallingCodes;

        var searchCriteria = name + ' ' + countryCallingCodes.join(' ');
        return new RegExp((0, _escapeStringRegexp2.default)(value.trim()), 'gi').test(searchCriteria);
      });
      this.setState({ filteredCountries: value === '' ? preferredCountries : filteredCountries, searchTerm: value, tabbedIndex: 0 });
    }
  }, {
    key: 'selectCountry',
    value: function selectCountry(country) {
      var _this4 = this;

      var mounted = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var multiSelect = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      var inputWithoutCountryCode = this.state.inputWithoutCountryCode;
      var countryCallingCodes = country.countryCallingCodes,
          alpha2 = country.alpha2;

      if (countryCallingCodes.length > 1 && !multiSelect) {
        return this.setState({ multiSelectOpen: true, multiSelectItem: country }, function () {
          _this4.multiSelect.style.zIndex = '101';
        });
      }
      var callingCode = multiSelect || countryCallingCodes[0];
      var phoneNumber = this.getFormattedNumber(callingCode, inputWithoutCountryCode);
      var validation = this.validateNumber(phoneNumber, multiSelect ? country.alpha2 : alpha2);
      this.setState({ selectedCountry: country, callingCode: callingCode, open: false, tabbedIndex: 0, phoneNumber: phoneNumber, validation: validation }, function () {
        _this4.cancelMultiSelect();
        _this4.onChangePhone(phoneNumber, mounted);
        if (!mounted) {
          _this4.phoneInput.focus();
        }
      });
    }
  }, {
    key: 'pageClick',
    value: function pageClick() {
      var _this5 = this;

      if (!this.mouseDownOnMenu) {
        this.setState({ open: false, tabbedIndex: 0 }, function () {
          _this5.countryDropdown.scrollTop = 0;
        });
        this.cancelMultiSelect();
      }
    }
  }, {
    key: 'onOpenHandler',
    value: function onOpenHandler() {
      var disabled = this.props.disabled;

      if (!disabled) {
        var open = this.state.open;

        this.setState({ open: !open });
        if (!open) {
          this.phoneInput.focus();
        }
      }
    }
  }, {
    key: 'clearInput',
    value: function clearInput() {
      var _state3 = this.state,
          open = _state3.open,
          selectedCountry = _state3.selectedCountry,
          callingCode = _state3.callingCode;

      if (open) {
        this.setState({ searchTerm: '', filteredCountries: this.getPreferredCountries(), multiSelectItem: [], multiSelectOpen: false });
      } else if (selectedCountry === 'unknown') {
        this.setState({ phoneNumber: '', inputWithoutCountryCode: '' });
        this.onChangePhone('');
      } else {
        var phoneNumber = this.getFormattedNumber(callingCode);
        this.setState({ phoneNumber: phoneNumber, inputWithoutCountryCode: '' });
        this.onChangePhone(phoneNumber);
        this.cancelMultiSelect();
      }
      this.phoneInput.focus();
    }
  }, {
    key: 'mouseDownHandler',
    value: function mouseDownHandler() {
      this.mouseDownOnMenu = true;
    }
  }, {
    key: 'mouseUpHandler',
    value: function mouseUpHandler() {
      this.mouseDownOnMenu = false;
    }
  }, {
    key: 'propChangeHandler',
    value: function propChangeHandler(props, mounted) {
      var _this6 = this;

      var selectedCountry = this.state.selectedCountry;
      var defaultCountry = props.defaultCountry,
          defaultValue = props.defaultValue;

      var countryNotSelected = Object.keys(selectedCountry).length < 1 && selectedCountry !== 'unknown';
      console.log(selectedCountry, countryNotSelected, defaultCountry, defaultValue);
      if (defaultValue) {
        var _validateNumber = this.validateNumber(defaultValue, 'unknown'),
            parsed = _validateNumber.parsed;

        if (parsed) {
          this.setState({ inputWithoutCountryCode: parsed.getNationalNumber() }, function () {
            _this6.selectCountry(_this6.lookupCountry(parsed.getCountryCode()), mounted);
          });
        } else {
          this.setState({ phoneNumber: defaultValue, inputWithoutCountryCode: defaultValue, selectedCountry: 'unknown' });
        }
      } else if (defaultCountry && countryNotSelected) {
        this.setState({ phoneNumber: '', inputWithoutCountryCode: '' }, function () {
          _this6.selectCountry(_countryData2.default.countries[defaultCountry], mounted);
        });
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var reset = nextProps.reset,
          defaultValue = nextProps.defaultValue;

      if (reset || this.props.defaultValue !== defaultValue) {
        this.propChangeHandler(nextProps, reset);
      }
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      window.addEventListener('mousedown', this._pageClick);
      this.propChangeHandler(this.props, true);
      this.setState({ filteredCountries: this.getPreferredCountries() });
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.removeEventListener('mousedown', this._pageClick);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this7 = this;

      var _state4 = this.state,
          open = _state4.open,
          selectedCountry = _state4.selectedCountry,
          phoneNumber = _state4.phoneNumber,
          filteredCountries = _state4.filteredCountries,
          searchTerm = _state4.searchTerm,
          paginateCount = _state4.paginateCount,
          inputWithoutCountryCode = _state4.inputWithoutCountryCode,
          multiSelectOpen = _state4.multiSelectOpen,
          multiSelectItem = _state4.multiSelectItem,
          lastPreferred = _state4.lastPreferred,
          tabbedIndex = _state4.tabbedIndex;
      var _props2 = this.props,
          noResultsMessage = _props2.noResultsMessage,
          className = _props2.className,
          removeToken = _props2.removeToken,
          paginate = _props2.paginate,
          paginateText = _props2.paginateText,
          placeholder = _props2.placeholder,
          maxHeight = _props2.maxHeight,
          disabled = _props2.disabled,
          containerClassName = _props2.containerClassName,
          inputClassName = _props2.inputClassName;
      var alpha2 = selectedCountry.alpha2;

      var flag = (this.missingFlags[alpha2] ? this.missingFlags[alpha2] : selectedCountry !== 'unknown' && Object.keys(selectedCountry).length > 0 && alpha2.toUpperCase()) || 'WW';
      return _react2.default.createElement(
        'div',
        { ref: function ref(input) {
            _this7.intlPhoneInput = input;
          }, className: (containerClassName ? containerClassName : '') + 'intl-phone-input' + (open ? ' open' : ''), onMouseDown: function onMouseDown() {
            return _this7.mouseDownHandler();
          }, onMouseUp: function onMouseUp() {
            return _this7.mouseUpHandler();
          } },
        _react2.default.createElement(
          'div',
          { className: 'input-group' },
          _react2.default.createElement(
            'div',
            { className: 'input-group-btn' },
            _react2.default.createElement(
              'button',
              { type: 'button', disabled: disabled, className: 'btn btn-secondary dropdown-toggle country-selector', onClick: function onClick(e) {
                  return _this7.onOpenHandler(e);
                } },
              flag && _react2.default.createElement(_FlagIcon2.default, { code: flag, size: 24, className: 'flag-icon' })
            )
          ),
          (open && searchTerm.length > 0 || !open && inputWithoutCountryCode.length > 0) && !disabled && _react2.default.createElement(
            'span',
            { className: 'remove-token-container' },
            _react2.default.createElement(
              'span',
              { style: { cursor: 'pointer' }, onClick: function onClick() {
                  return _this7.clearInput();
                } },
              removeToken
            )
          ),
          _react2.default.createElement('input', { type: 'text',
            ref: function ref(input) {
              _this7.phoneInput = input;
            },
            className: 'form-control phone-input' + (inputClassName ? inputClassName : ''),
            onKeyDown: function onKeyDown(e) {
              return _this7.onKeyDown(e);
            },
            placeholder: open ? placeholder : '',
            value: open ? searchTerm : phoneNumber,
            disabled: disabled,
            onChange: function onChange(e) {
              return open ? _this7.onChangeTypeAhead(e.target.value) : _this7.onChangePhone(e.target.value);
            }
          })
        ),
        _react2.default.createElement(
          'div',
          { ref: function ref(dropdown) {
              _this7.countryDropdown = dropdown;
            }, className: 'dropdown-menu country-dropdown', style: { maxHeight: open ? maxHeight : 0 } },
          filteredCountries && filteredCountries.length > 0 && filteredCountries.map(function (country, index) {
            var name = country.name,
                alpha2 = country.alpha2,
                countryCallingCodes = country.countryCallingCodes;

            var paginateTo = paginate && parseInt(paginate) * paginateCount;
            if (index <= paginateTo) {
              return _react2.default.createElement(
                'div',
                { className: 'dropdown-item' + (lastPreferred && lastPreferred.alpha2 === alpha2 && searchTerm === '' ? ' preferred' : '') + (alpha2 === selectedCountry.alpha2 ? ' selected' : '') + (tabbedIndex === index + 1 ? ' tabbed' : ''),
                  key: alpha2 + '-' + index,
                  onClick: function onClick() {
                    return _this7.selectCountry(country);
                  } },
                _react2.default.createElement(_FlagIcon2.default, { code: _this7.missingFlags[alpha2] ? _this7.missingFlags[alpha2] : alpha2, size: 30, className: 'flag-icon' }),
                name,
                ' ',
                countryCallingCodes.map(function (code, index) {
                  return _react2.default.createElement(
                    'span',
                    { className: 'country-code', key: code },
                    code,
                    index !== countryCallingCodes.length - 1 && _react2.default.createElement('span', { key: code + '-divider', className: 'country-code-divider' })
                  );
                })
              );
            }
            if (index - 1 === paginateTo) {
              return _react2.default.createElement(
                'div',
                { className: 'dropdown-item', key: 'addit-results-' + index, onClick: function onClick() {
                    return _this7.setState({ paginateCount: paginateCount + 1 });
                  } },
                paginateText
              );
            }
          }),
          filteredCountries && filteredCountries.length === 0 && _react2.default.createElement(
            'div',
            { className: 'dropdown-item' },
            noResultsMessage
          ),
          _react2.default.createElement(
            'div',
            { ref: function ref(select) {
                _this7.multiSelect = select;
              }, className: 'text-center calling-code-multi-select' + (multiSelectOpen ? ' open' : '') },
            _react2.default.createElement(
              'button',
              { type: 'button', 'aria-label': 'Close', onClick: function onClick() {
                  return _this7.cancelMultiSelect();
                }, className: 'btn btn-outline btn-outline-danger multi-select-back-btn' },
              'Close'
            ),
            Object.keys(multiSelectItem).length > 0 && multiSelectItem.countryCallingCodes.map(function (item) {
              return _react2.default.createElement(
                'button',
                { key: item, type: 'button', onClick: function onClick() {
                    return _this7.selectCountry(multiSelectItem, false, item);
                  }, className: 'btn btn-secondary country-btn' },
                item
              );
            })
          )
        )
      );
    }
  }]);

  return IntlTelInput;
}(_react.Component);

exports.default = IntlTelInput;


IntlTelInput.defaultProps = {
  removeToken: _react2.default.createElement(
    'span',
    null,
    '\xD7'
  ),
  noResultsMessage: 'No results available',
  paginateText: 'Display additional results...',
  paginate: 50,
  placeholder: 'Start typing to search...',
  maxHeight: 300,
  defaultCountry: 'US',
  minLengthMessage: 'Too short to be a valid number',
  maxLengthMessage: 'Too long to be a valid number',
  callingCodeMessage: 'Please select a country code',
  catchAllMessage: 'Not a valid number'
};

IntlTelInput.propTypes = {
  removeToken: _react.PropTypes.oneOfType([_react.PropTypes.element, _react.PropTypes.string]),
  preferredCountries: _react.PropTypes.arrayOf(_react.PropTypes.string),
  defaultValue: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]),
  noResultsMessage: _react.PropTypes.string,
  paginateText: _react.PropTypes.string,
  paginate: _react.PropTypes.number,
  disabled: _react.PropTypes.bool,
  placeholder: _react.PropTypes.string,
  maxHeight: _react.PropTypes.number,
  defaultCountry: _react.PropTypes.string,
  onChange: _react.PropTypes.func,
  minLengthMessage: _react.PropTypes.string,
  maxLengthMessage: _react.PropTypes.string,
  callingCodeMessage: _react.PropTypes.string,
  catchAllMessage: _react.PropTypes.string,
  containerClassName: _react.PropTypes.string,
  inputClassName: _react.PropTypes.string
};
